common/solver/cb_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,1593 @@
+        krylov_bases(krylov_idx, krylov_row, rhs) = zero<ValueType>();
+    }
+
+    if (residual_row < krylov_dim + 1 && residual_col < num_rhs) {
+        residual_norm_collection[residual_row * stride_residual_nc +
+                                 residual_col] = zero<ValueType>();
+    }
+}
+
+
+// Must be called with at least `num_rows * num_rhs` threads in total.
+template <size_type block_size, typename ValueType, typename Accessor3d>
+__global__ __launch_bounds__(block_size) void initialize_2_2_kernel(
+    size_type num_rows, size_type num_rhs,
+    const ValueType *__restrict__ residual, size_type stride_residual,
+    const remove_complex<ValueType> *__restrict__ residual_norm,
+    ValueType *__restrict__ residual_norm_collection, Accessor3d krylov_bases,
+    ValueType *__restrict__ next_krylov_basis, size_type stride_next_krylov,
+    size_type *__restrict__ final_iter_nums)
+{
+    const auto global_id = thread::get_thread_id_flat();
+    const auto krylov_stride =
+        helper_functions_accessor<Accessor3d>::get_stride(krylov_bases);
+    const auto row_idx = global_id / krylov_stride[1];
+    const auto col_idx = global_id % krylov_stride[1];
+
+    if (global_id < num_rhs) {
+        residual_norm_collection[global_id] = residual_norm[global_id];
+        final_iter_nums[global_id] = 0;
+    }
+
+    if (row_idx < num_rows && col_idx < num_rhs) {
+        auto value = residual[row_idx * stride_residual + col_idx] /
+                     residual_norm[col_idx];
+        krylov_bases(0, row_idx, col_idx) = value;
+        next_krylov_basis[row_idx * stride_next_krylov + col_idx] = value;
+    }
+}
+
+
+__global__
+    __launch_bounds__(default_block_size) void increase_final_iteration_numbers_kernel(
+        size_type *__restrict__ final_iter_nums,
+        const stopping_status *__restrict__ stop_status, size_type total_number)
+{
+    const auto global_id = thread::get_thread_id_flat();
+    if (global_id < total_number) {
+        final_iter_nums[global_id] += !stop_status[global_id].has_stopped();
+    }
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_dot_size) void multinorm2_kernel(
+    size_type num_rows, size_type num_cols,
+    const ValueType *__restrict__ next_krylov_basis,
+    size_type stride_next_krylov, remove_complex<ValueType> *__restrict__ norms,
+    const stopping_status *__restrict__ stop_status)
+{
+    const auto tidx = threadIdx.x;
+    const auto tidy = threadIdx.y;
+    const auto col_idx = blockIdx.x * default_dot_dim + tidx;
+    const auto num = ceildiv(num_rows, gridDim.y);
+    const auto start_row = blockIdx.y * num;
+    const auto end_row =
+        ((blockIdx.y + 1) * num > num_rows) ? num_rows : (blockIdx.y + 1) * num;
+    // Used that way to get around dynamic initialization warning and
+    // template error when using `reduction_helper_array` directly in `reduce`
+    __shared__ UninitializedArray<remove_complex<ValueType>,
```

Comment:
Maybe also store the `rc_vtype` for this function, this could improve coding style a bit.